File: lib-segment/src/main/java/websiteschema/mpsegment/dict/IWordArray.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package websiteschema.mpsegment.dict;

import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;


/**
 * @author twer
 */
public interface IWordArray<Word extends IWord> {

    public Word find(String word);

    public Word[] getWordItems();

    public void add(Word word);
}


class BinaryWordArray<Word extends IWord> implements IWordArray<Word> {

    IWord[] wordItems;

    BinaryWordArray(Word[] wordItems) {
        this.wordItems = wordItems;
    }

    @Override
    public Word find(String word) {
        int index = lookupWordItem(word);
        if (index >= 0) {
            return (Word)wordItems[index];
        }
        return null;
    }

    @Override
    public void add(Word word) {
        IWord[] temp = new IWord[wordItems.length + 1];
        System.arraycopy(wordItems, 0, temp, 0, wordItems.length);
        temp[wordItems.length] = word;
        Arrays.sort(temp, new Comparator<IWord>() {
            @Override
            public int compare(IWord o1, IWord o2) {
                return o1.getWordName().compareTo(o2.getWordName());
            }
        });
        wordItems = temp;
    }

    private int lookupWordItem(String word) {
        int left = 0;
        for (int right = wordItems.length - 1; left <= right;) {
            int mid = (left + right) / 2;
            int comp = wordItems[mid].getWordName().compareTo(word);
            if (comp == 0) {
                return mid;
            }
            if (comp < 0) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }

    @Override
    public Word[] getWordItems() {
        return (Word[])wordItems;
    }
}

class HashWordArray<Word extends IWord> implements IWordArray<Word> {

    HashMap<String, Integer> wordIndex;
    IWord[] wordItems;

    HashWordArray(Word[] words) {
        this.wordIndex = new HashMap<String, Integer>(words.length);
        int i = 0;
        for (Word word : words) {
            wordIndex.put(word.getWordName(), i++);
        }

        wordItems = words;
    }

    @Override
    public Word find(String word) {
        Integer index = wordIndex.get(word);
        if (null != index && index >= 0) {
            return (Word)wordItems[index];
        }
        return null;
    }

    @Override
    public Word[] getWordItems() {
        return (Word[])wordItems;
    }

    @Override
    public void add(Word word) {
        IWord[] temp = new IWord[wordItems.length + 1];
        System.arraycopy(wordItems, 0, temp, 0, wordItems.length);
        int index = wordItems.length;
        temp[index] = word;
        wordIndex.put(word.getWordName(), index);
        wordItems = temp;
    }
}

File: lib-segment/src/main/scala/websiteschema/mpsegment/dict/IWordArray.scala
package websiteschema.mpsegment.dict

import scala.collection.mutable.Map
import scala.collection.mutable.OpenHashMap
import scala.collection.mutable

trait IWordArray {

  def find(word: String): IWord

  def getWordItems(): Array[IWord]

  def add(word: IWord)
}

object BinaryWordArray {

  val ordering: Ordering[IWord] = Ordering[String].on[IWord](_.getWordName())

  def apply(wordItems: Array[IWord]) = {
    val wordArray = new BinaryWordArray()
    wordArray.wordSet = new mutable.TreeSet[IWord]()(ordering)
    wordArray.wordSet ++= wordItems
    wordArray
  }
}

class BinaryWordArray extends IWordArray {

  var wordSet = new mutable.TreeSet[IWord]()(BinaryWordArray.ordering)

  override def find(word: String): IWord =
    wordSet.find(w => w.getWordName().equals(word)) match {
      case Some(w) => w
      case _ => null
    }

  override def add(word: IWord) {
    wordSet.add(word)
  }

  override def getWordItems(): Array[IWord] = wordSet.toArray
}

object HashWordArray {
  def apply(words: Array[IWord]) = {
    val wordArray = new HashWordArray()
    wordArray.wordIndex = OpenHashMap[String, IWord]()
    for (word <- words) {
      wordArray.wordIndex += (word.getWordName() -> word)
    }

    wordArray
  }
}

class HashWordArray extends IWordArray {

  var wordIndex: Map[String, IWord] = null
  var wordItems: Array[IWord] = null

  override def find(word: String): IWord = {
    wordIndex.get(word) match {
      case Some(w) => w
      case _ => null
    }
  }

  override def getWordItems(): Array[IWord] = {
    return wordIndex.toList.map(_._2).toArray
  }

  override def add(word: IWord) {
    wordIndex += (word.getWordName() -> word)
  }
}

